cmd/strace.go

```diff
@@ -76,22 +77,96 @@ func strace(c *cli.Context) error {
 			return nil
 		}
 
-		// Print register value on every second trap so that we print only after the syscall is complete
+		// Print register values on every second trap so that we print only after the syscall is complete
 		if traps > 0 && traps%2 == 0 {
-			origRax, err := readOrigRax(childPid)
+			var regs syscall.PtraceRegs
+			err = syscall.PtraceGetRegs(childPid, &regs)
 			if err != nil {
 				return fmt.Errorf("error while reading Orig_rax for child pid %d: %w ", childPid, err)
 			}
-			log.WithFields(log.Fields{
-				"orig_rax": origRax,
-				"syscall":  syscallNameByRax[origRax],
-			}).Info("Caught syscall")
+
+			// Syscall-related data is stored in a number of registers, namely:
+			// - rax (system call number)
+			// - rdi (first arg)
+			// - rsi (second arg)
+			// - rdx (third arg)
+			// - r10 (fourth arg)
+			// - r8 (fifth arg)
+			// - r9 (sixth arg)
+			// For each syscall, we need to hardcode which arguments are of interest
+			// Hence, for educational purposes, only selected syscalls have been
+			// defined (i.e. are recognized) below. The list can grow over time.
+			syscallName := syscallNameByRax[regs.Orig_rax]
+
+			switch syscallName {
+
+			case "openat":
+				dirfd := regs.Rdi
+				path, err := peekText(childPid, uintptr(regs.Rsi))
+				flags := regs.Rdx
+				if err != nil {
+					return fmt.Errorf("error while reading path text for child pid %d: %w", childPid, err)
+				}
+				fmt.Printf("openat(%d, '%s', %d)\n", dirfd, string(path), flags)
+			case "open":
+				path, err := peekText(childPid, uintptr(regs.Rdi))
+				if err != nil {
+					return fmt.Errorf("error while reading path text for child pid %d: %w", childPid, err)
+				}
+				flags := regs.Rsi
+				fmt.Printf("open('%s', %d)\n", path, flags)
+			case "read":
+				fd := regs.Rdi
+				buf := regs.Rsi
+				count := regs.Rdx
+				fmt.Printf("read(%d, %d, %d)\n", fd, buf, count)
+			case "write":
+				fd := regs.Rdi
+				text, err := peekText(childPid, uintptr(regs.Rsi))
+				if err != nil {
+					return fmt.Errorf("error while reading path text for child pid %d: %w", childPid, err)
+				}
+				nbytes := regs.Rdx
+				fmt.Printf("write(%d, '%s', %d)\n", fd, text, nbytes)
+			case "close":
+				fd := regs.Rdi
+				fmt.Printf("close(%d)\n", fd)
+			default:
+				// Print out syscall(n/a) as indication that we cannot parse paramters
+				fmt.Printf("%s(n/a)\n", syscallName)
+			}
+
 		}
 
 		traps++
 	}
 }
 
+// Reads text from another process' memory
+func peekText(pid int, addr uintptr) (string, error) {
+	var sb strings.Builder
+
+	// We will read text bytes one by one
+	// since we don't know the length
+	buf := make([]byte, 1)
+	offset := uintptr(0)
+	for {
+		_, err := syscall.PtracePeekText(pid, addr+offset, buf)
```

Comment:
Can't we do better than this? e.g. call `syscall.PtracePeekText` once, read the full buffer (until we encounter `\0`) and call it again if we reached the end? We are doing as many `syscall.PtracePeekText` calls as we have bytes to read, which sounds suboptimal.

Reply:
True, this not a great implementation but my quick analysis of https://golang.org/src/syscall/syscall_linux.go (and some manual testing of the behavior) shows that basically the way PtracePeekText works is that it just starts reading at the address given and reads as many bytes as the provided buffer can hold. If I give it a bigger buffer e.g. of length 1024, it will read 1024 bytes... which means it will read the string, go past the null byte and read some other random data from the memory. In the worst case we could get SEGV if we read like that. Unfotunately we don't know the string length in advance so we are blind here and need to read carefully. 

I don't think that any optimization is worth investing time in this code but I guess we could alternatively try increasing the buffer length in a manner similar to how TCP deals with windows i.e. grow the buffer exponentially with each iteration and reduce its size on SEGV (we would need to handle SEGV signalling so that we don't die). I will see if it works out of curiosity.

Reply:
Ok, so actually if you try to read too much, ptrace will return an I/O error. So we would need to react on these errors and try ptracing with different buffer sizes. Doable but I guess it doesn't make much sense to implement it in this code?